src/database.rs

```diff
@@ -219,17 +255,46 @@ impl Handler<RegisterHash> for DatabaseManager {
 
     fn handle(&mut self, msg: RegisterHash, _ctx: &mut Self::Context) -> Self::Result {
         let map_hash = crate::filemap::hash_bundles(msg.files.iter().map(|(map, _path)| map));
-        let desc = FileDesc {
+        let reporter = msg.reporter;
+        let desc = Arc::new(FileDesc {
             map_hash,
             files: msg.files,
             inline_data: msg.inline_data,
-            valid_to: msg.valid_to,
-        };
-        self.files.insert(map_hash, Arc::new(desc));
+            valid_to: msg.valid_to.clone(),
+        });
+        if let Some((prev, prev_reporter)) = self.files.insert(map_hash, (desc.clone(), reporter)) {
```

Comment:
Wouldn't a db read + write be cheaper than 2 writes, in a pessimistic case?

Reply:
it is my java habit. write is cheaper than hash lookup so pesimistic lookup + write + lookup + write is almost as slow as lookup + lookup + write. But in RUST maybe this is better:

```rust
match self.files.entry(map_hash) {
  Entry::Occupied(mut ent) => {
    let prev_ent = ent.get_mut();
    ...
    if !old_is_longer {
      prev_ent.0 = desc.clone();
      desc.log_event("share extend");
     }
    }
    Entry::Vacant(ent) => {
      ...
    }
}